### PR TITLE
Improve CDCImporter process FinishedRecord, and invalidate cache after acked

### PR DESCRIPTION
**Pull request by azexcy**
Related: #22500 

Changes proposed in this pull request:
  - Improve CDCImporter finished flag
  - Invalidate cache after acked

---

Before committing this PR, I'm sure that I have checked the following options:
- [ ] My code follows the [code of conduct](https://shardingsphere.apache.org/community/en/involved/conduct/code/) of this project.
- [ ] I have self-reviewed the commit code.
- [ ] I have (or in comment I request) added corresponding labels for the pull request.
- [ ] I have passed maven check locally : `./mvnw clean install -B -T1C -Dmaven.javadoc.skip -Dmaven.jacoco.skip -e`.
- [ ] I have made corresponding changes to the documentation.
- [ ] I have added corresponding unit tests for my changes.

### CODE REVIEW

File: kernel/data-pipeline/scenario/cdc/core/src/main/java/org/apache/shardingsphere/data/pipeline/cdc/core/importer/CDCImporter.java [hunk 1]
```diff
@@ -86,7 +86,7 @@ protected void runBlocking() {
             } else {
                 doWithoutSorting(channelProgressPairs);
             }
-            if (channelProgressPairs.isEmpty()) {
+            if (channelProgressPairs.isEmpty() && ackCache.estimatedSize() == 0) {
                 break;
             }
```
Comment (sandynz) [inline kernel/data-pipeline/scenario/cdc/core/src/main/java/org/apache/shardingsphere/data/pipeline/cdc/core/importer/CDCImporter.java:91]: 1, Client could have related re-subscription

2, If client always no ack, then server side could close streaming after some time out

**Review by sandynz — CHANGES_REQUESTED** [review]

File: kernel/data-pipeline/scenario/cdc/client/src/main/java/org/apache/shardingsphere/data/pipeline/cdc/client/handler/RetryStreamingExceptionHandler.java [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+public class RetryStreamingExceptionHandler implements ExceptionHandler {
+    
+    private final AtomicInteger retryTimes = new AtomicInteger(0);
+    
+    private CDCClient cdcClient;
+    
+    @Override
+    public void setCDCClient(final CDCClient cdcClient) {
+        this.cdcClient = cdcClient;
+    }
```
Comment (sandynz) [inline kernel/data-pipeline/scenario/cdc/client/src/main/java/org/apache/shardingsphere/data/pipeline/cdc/client/handler/RetryStreamingExceptionHandler.java:46]: Could be replaced by `@Setter`

File: kernel/data-pipeline/scenario/cdc/client/src/main/java/org/apache/shardingsphere/data/pipeline/cdc/client/handler/RetryStreamingExceptionHandler.java [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+            return;
+        }
+        ClientConnectionContext connectionContext = ctx.channel().attr(ClientConnectionContext.CONTEXT_KEY).get();
+        if (retryTimes.get() > 5) {
+            log.warn("Retry times exceed 5, stop streaming");
+            connectionContext.getStreamingIds().forEach(each -> CompletableFuture.runAsync(() -> cdcClient.stopStreaming(each)));
+            return;
+        }
+        retryTimes.incrementAndGet();
+        TimeUnit.SECONDS.sleep(Math.min(5 >> retryTimes.get(), 10));
```
Comment (sandynz) [inline kernel/data-pipeline/scenario/cdc/client/src/main/java/org/apache/shardingsphere/data/pipeline/cdc/client/handler/RetryStreamingExceptionHandler.java:68]: It's better to extract retryTimes and sleepTime as fields, so users could customize them

File: kernel/data-pipeline/scenario/cdc/client/src/main/java/org/apache/shardingsphere/data/pipeline/cdc/client/handler/RetryStreamingExceptionHandler.java [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+    
+    @Override
+    public void setCDCClient(final CDCClient cdcClient) {
+        this.cdcClient = cdcClient;
+    }
+    
+    @Override
+    public void handleServerException(final ChannelHandlerContext ctx, final ServerErrorResult result) {
+        // Server exception maybe not be resolved by retrying
+        log.error("Server error, code: {}, message: {}", result.getErrorCode(), result.getErrorMessage());
```
Comment (sandynz) [inline kernel/data-pipeline/scenario/cdc/client/src/main/java/org/apache/shardingsphere/data/pipeline/cdc/client/handler/RetryStreamingExceptionHandler.java:51]: Looks it's not defined clearly, users might need to customize it again, but could not just reuse RetryStreamingExceptionHandler.

Could we extract retry logic to utility class?

Or extract two handlers, one is exception handler, and another is server error handler?

**Review by sandynz — CHANGES_REQUESTED** [review]

File: kernel/data-pipeline/scenario/cdc/client/src/test/java/org/apache/shardingsphere/data/pipeline/cdc/client/example/Bootstrap.java [hunk 5]
```diff
@@ -41,14 +42,14 @@ public static void main(final String[] args) {
...
         String address = "127.0.0.1";
-        CDCClientConfiguration clientConfig = new CDCClientConfiguration(address, 33071, records -> log.info("records: {}", records), new RetryStreamingExceptionHandler(3, 5000));
         int reconnectCount = 0;
         int maxReconnectCount = 5;
         while (reconnectCount < maxReconnectCount) {
-            try (CDCClient cdcClient = new CDCClient(clientConfig)) {
-                cdcClient.connect();
+            try (CDCClient cdcClient = new CDCClient(new CDCClientConfiguration(address, 33071, 5000))) {
+                LoggerExceptionErrorHandler loggerExceptionErrorHandler = new LoggerExceptionErrorHandler();
+                cdcClient.connect(records -> log.info("records: {}", records), new RetryStreamingExceptionHandler(cdcClient, 5, 5000), loggerExceptionErrorHandler);
```
Comment (sandynz) [inline kernel/data-pipeline/scenario/cdc/client/src/test/java/org/apache/shardingsphere/data/pipeline/cdc/client/example/Bootstrap.java:50]: `reconnectCount` is different with `Retry...Handler`, `catch (final Exception ex)` could be improved

File: test/e2e/operation/pipeline/src/test/java/org/apache/shardingsphere/test/e2e/data/pipeline/cases/cdc/CDCE2EIT.java [hunk 6]
```diff
@@ -169,10 +169,10 @@ private DataSource createStandardDataSource(final PipelineContainerComposer cont
     private void startCDCClient(final PipelineContainerComposer containerComposer, final DialectDatabaseMetaData dialectDatabaseMetaData) {
         DataSource dataSource = createStandardDataSource(containerComposer, PipelineContainerComposer.DS_4);
         DataSourceRecordConsumer recordConsumer = new DataSourceRecordConsumer(dataSource, containerComposer.getDatabaseType());
-        CDCClientConfiguration cdcConfig = new CDCClientConfiguration("localhost", containerComposer.getContainerComposer().getProxyCDCPort(), recordConsumer, new LoggerExceptionHandler());
         String schema = dialectDatabaseMetaData.isSchemaAvailable() ? "test" : "";
-        CDCClient cdcClient = new CDCClient(cdcConfig);
-        cdcClient.connect();
+        CDCClient cdcClient = new CDCClient(new CDCClientConfiguration("localhost", containerComposer.getContainerComposer().getProxyCDCPort(), 5000));
+        LoggerExceptionErrorHandler handler = new LoggerExceptionErrorHandler();
+        cdcClient.connect(recordConsumer, handler, handler);
```
Comment (sandynz) [inline test/e2e/operation/pipeline/src/test/java/org/apache/shardingsphere/test/e2e/data/pipeline/cases/cdc/CDCE2EIT.java:175]: Could we improve `handler, handler` parameter with closure?

**Review by sandynz — CHANGES_REQUESTED** [review]

File: kernel/data-pipeline/scenario/cdc/client/src/test/java/org/apache/shardingsphere/data/pipeline/cdc/client/example/Bootstrap.java (hunk 5 above)
Comment (azexcy) [reply kernel/data-pipeline/scenario/cdc/client/src/test/java/org/apache/shardingsphere/data/pipeline/cdc/client/example/Bootstrap.java]: Remove `reconnect` now, it's not generic logic

**Review by sandynz — APPROVED** [review]

**Pull request merged**